pkg/streamingpromql/optimize/ast/collapse_constants_test.go

```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ast
+
+import (
+	"context"
+	"testing"
+
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCollapseConstants(t *testing.T) {
+	testCases := map[string]string{
+		"2":                    "2",
+		"-2":                   "-2",
+		"-(2)":                 "-2",
+		"+(2)":                 "2",
+		"2 + 3":                "5",
+		"(2 + 3)":              "5",
+		"2 - 3":                "-1",
+		"2 * 3":                "6",
+		"3 / 2":                "1.5",
+		"5 % 2":                "1",
+		"2 ^ 3":                "8",
+		"1 atan2 3":            "0.3217505543966422",
+		"abs(vector(2 + 3))":   "abs(vector(5))",
+		"vector(2 + 3)":        "vector(5)",
+		"vector(2 + 3)[1m:1m]": "vector(5)[1m:1m]",
+		"sum(vector(2 + 3))":   "sum(vector(5))",
+		"topk(2 + 3, metric)":  "topk(5, metric)",
+
+		"foo * 2":    "foo * 2",
+		"2 * foo":    "2 * foo",
+		"bar * foo":  "bar * foo",
+		"2 > bool 3": "2 > bool 3",
+
+		"metric":        "metric",
+		"(metric)":      "(metric)",
+		"-metric":       "-metric",
+		"metric[1m]":    "metric[1m]",
+		"metric[1m:1m]": "metric[1m:1m]",
+		`"abc"`:         `"abc"`,
+	}
```

Comment:
(nit) would be good to have some cases where constants are a small part of a larger query with lots of irrelevant parts.

Reply:
Could you give an example of the kind of thing you had in mind? There are already a couple of cases like this earlier in this list (eg. `sum(vector(2 + 3))`)

Reply:
Just some needlessly complex query like `label_join(...) + (sum(vector(2+3)) / (sum(count) + sum(min(series))) * 600 + 3000`, to check the `2+3` gets collapsed and nothing else is unexpectedly changed.

Also, not sure if it's already covered, but that example could be interesting where we want to test the order of operations. ie we don't want the last part to become `* 3600`.

Reply:
Done: https://github.com/grafana/mimir/pull/11124/commits/ce15a445676542222d8482f6525f4f06a83a6a3a